### DISCUSSION TITLE
Invoking functions using HTTP

### DISCUSSION THREAD

**Discussion by faromero** [vhive-serverless/vHive#336] [Q&A] [answered]
Do you have any examples of how to invoke functions using HTTP? Or any guidance on how that can be done?

Thanks!

**Comment by shyamjesal** [comment, marked as answer]
Once the cluster is up and running like mentioned in [this](https://github.com/ease-lab/vhive/blob/main/docs/quickstart_guide.md#iii-setup-a-single-node-cluster) step, you can use knative client to deploy functions.
```bash
kn service apply helloworld -f vhive/configs/knative_workloads/helloworld.yaml
Creating service 'helloworld' in namespace 'default':

  0.056s The Route is still working to reflect the latest desired specification.
  0.100s ...
  0.160s Configuration "helloworld" is waiting for a Revision to become ready.
591.046s ...
591.084s Ingress has not yet been reconciled.
591.180s Waiting for load balancer to be ready
591.398s Ready to serve.

Service 'helloworld' created to latest revision 'helloworld-00001' is available at URL:
http://helloworld.default.192.168.1.240.sslip.io
```
The `url` on the last line is the HTTP endpoint which can be used to invoke the function.

**Reply by faromero** [reply]
Ok great, thank you! And does invoking via HTTP support passing arguments in the typical HTTP fashion (e.g., after a ?) that are accessible from inside the function. With gRPC this can be done with the [function stub](https://github.com/ease-lab/vhive/blob/main/examples/invoker/client.go#L176-L182).

**Reply by shyamjesal** [reply]
The HTTP request is directly forwarded as-is by the Knative queue-proxy to the firecracker VM. I believe the HTTP arguments should work right away, but I haven't tried it yet. You can test this using a simple example like one of [this](https://github.com/knative/docs/tree/mkdocs/docs/serving/samples/hello-world).
You will need to modify the knative yaml file for vhive. [Here](https://github.com/ease-lab/vhive/blob/main/configs/knative_workloads/helloworld.yaml) is an example for simple gRPC helloworld.

**Reply by ustiugov** [reply]
@faromero note that gRPC is incompatible with pure HTTP clients/servers. You need to choose one of them. 

any reason why you would prefer HTTP over gRPC?

**Reply by faromero** [reply]
It's easier to make requests from _gg_ using HTTP than using gRPC. Alternatively, we were thinking of putting a simple proxy HTTP server that simply "translates" the request from _gg_ to gRPC, and from there sends it to vHive as the examples do (and vice-versa).

**Reply by ustiugov** [reply]
@faromero I see. in this case, I recommend you to use HTTP everywhere (clients and servers) that would not need any modifications in vHive, except slightly different knative yaml files that you would have to use (same as in the Knative examples, to which Shyam gave a link). once you show us your code (best by submitting a PR to this repo), we can see if we want to move from HTTP to gRPC later on.

Adding a HTTP/gRPC translation layer is definitely not a great idea because it would come with measurable overheads and also would be difficult to maintain as an extra vHive component

**Reply by niravnshah** [reply]
Hi @shyamjesal,
while trying to invoke the function using http, i am getting following error. Any idea what might be going wrong?

```
nirav@984fee0096e3:~/vhive$ kn service apply myserv -f configs/knative_workloads/pyaes.yaml
Creating service 'myserv' in namespace 'default':

  0.262s The Route is still working to reflect the latest desired specification.
  0.313s ...
  0.364s Configuration "myserv" is waiting for a Revision to become ready.
 10.254s ...
 10.302s Ingress has not yet been reconciled.
 10.413s Waiting for load balancer to be ready
 10.618s Ready to serve.

Service 'myserv' created to latest revision 'myserv-00001' is available at URL:
http://myserv.default.192.168.1.240.sslip.io
nirav@984fee0096e3:~/vhive$ kn service list
NAME           URL                                                  LATEST               AGE    CONDITIONS   READY   REASON
helloworld-0   http://helloworld-0.default.192.168.1.240.sslip.io   helloworld-0-00001   108m   3 OK / 3     True
myserv         http://myserv.default.192.168.1.240.sslip.io         myserv-00001         25s    3 OK / 3     True
nirav@984fee0096e3:~/vhive$ curl http://myserv.default.192.168.1.240.sslip.io
stream error: stream ID 5; INTERNAL_ERROR; received from peer
```

**Reply by ustiugov** [reply]
pyaes and other functions in that folder support gRPC only.

**Reply by niravnshah** [reply]
Found a way to invoke the gRPC function thru command line.
 
`$grpcurl -d '{"name": "Nirav", "vHiveMetadata": "1234"}' -import-path . -proto ./vhive/examples/invoker/helloworld.proto -plaintext helloworld-0.default.192.168.1.240.sslip.io:80 helloworld.Greeter/SayHello
`
This gives output as...
`{
  "message": "Hello, Nirav!"
}
`

This helps in determining if the installed services are running fine or not.
